### PR TITLE
Defer evaluation of api and app keys + read from `run_state`

### PR DESCRIPTION
**Pull request by olivielpeau**
Based on #345 (thanks @mfischer), rebased on top of master, and minus the changes on the `compile_time` property in `dd-handler`. Details in the commit messages and in #345.

Tested on both Chef 11 and 12, seems to work fine.

Related to #377

### CODE REVIEW

**Comment by olivielpeau** [comment]
Hmm actually I realize now that enabling the datadog handler at compile time could defeat the purpose of pulling the keys from the `run_state`. Unless other recipes set the `run_state` at compile time before `dd-handler` is compiled the `run_state` won't be populated with the keys.

We could add an attribute to only enable the handler at converge time.

File: libraries/recipe_helpers.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+        run_state_or_attribute(node, 'api_key')
+      end
+
+      def application_key(node)
+        run_state_or_attribute(node, 'application_key')
+      end
+
+      private
+
+      def run_state_or_attribute(node, attr)
```
Comment (degemer) [inline libraries/recipe_helpers.rb]: Should we rename `attr`? It's already used by Ruby (not in this context though) - http://ruby-doc.org/core-2.0.0/Module.html#method-i-attr

File: README.md [hunk 2]
```diff
@@ -86,10 +86,16 @@ Usage
 
 1. Add this cookbook to your Chef Server, either by installing with knife or by adding it to your Berksfile:
   ```
-  cookbook 'datadog', '~> 2.1.0'
+  cookbook 'datadog', '~> 2.7.0'
   ```
-2. Add your API Key as a node attribute via an `environment` or `role` or by declaring it in another cookbook at a higher precedence level.
-3. Create an 'application key' for `chef_handler` [here](https://app.datadoghq.com/account/settings#api), and add it as a node attribute, as in Step #2.
+2. Add your API Key either:
```
Comment (degemer) [inline README.md]: Should we add at the top of the README that these instructions are not for 2.7.0?

File: recipes/dd-agent.rb [hunk 3]
```diff
@@ -51,27 +51,25 @@
 # To add integration-specific configurations, add 'datadog::config_name' to
 # the node's run_list and set the relevant attributes
 #
-raise "Add a ['datadog']['api_key'] attribute to configure this node's Datadog Agent." if node['datadog'] && node['datadog']['api_key'].nil?
```
Comment (degemer) [inline recipes/dd-agent.rb]: Should we keep it? We'd have to change the message. "Add a ['datadog']['api_key'] attribute or set the ['datadog']['api_key'] run_state to configure this node's Datadog Agent."

File: templates/default/datadog.conf.erb [hunk 4]
```diff
@@ -141,11 +141,13 @@ if node['datadog']['legacy_integrations']
 end
 -%>
 
-<% if ! @extra_config.empty? -%>
+<% if ! node['datadog']['extra_config'].empty? -%>
 # ========================================================================== #
 # Other config options
 # ========================================================================== #
-  <% @extra_config.each do |k, v| -%>
+  <% @node['datadog']['extra_config'].each do |k, v| -%>
```
Comment (degemer) [inline templates/default/datadog.conf.erb]: `@`?

Comment (degemer) [inline templates/default/datadog.conf.erb]: Weird it's wasn't caught by the tests...

File: recipes/dd-agent.rb (hunk 3 above)
Comment (olivielpeau) [inline recipes/dd-agent.rb]: To keep it we'd need to evaluate it at execution time. I'll try that.

File: templates/default/datadog.conf.erb (hunk 4 above)
Comment (olivielpeau) [inline templates/default/datadog.conf.erb]: ha, good catch, apparently `@node` also works... Changing this to `node` anyway

File: README.md (hunk 2 above)
Comment (olivielpeau) [inline README.md]: Done

File: recipes/dd-agent.rb [hunk 5]
```diff
@@ -17,6 +17,13 @@
 # limitations under the License.
 #
 
+# Fail here at converge time if no api_key is set
+ruby_block 'datadog-api-key-set' do
+  block do
+    raise "Set ['datadog']['api_key'] as an attribute or on the node's run_state to configure this node's Datadog Agent." if Chef::Datadog.api_key(node).nil?
```
Comment (degemer) [inline recipes/dd-agent.rb]: 🍰

**Pull request merged**